events/events.go

```diff
@@ -11,92 +13,106 @@ type Event interface{}
 // EmitterInterface Root interface for events dispatch
 type EmitterInterface interface {
 	// Emit Sends an event to the subscribed listeners
-	Emit(Event)
+	Emit(context.Context, Event)
 
-	// Subscribe Registers a callback that will receive emitted events
-	Subscribe(context.Context, func(Event))
+	// Subscribe Returns a channel that receives emitted events
+	Subscribe(context.Context) <-chan Event
 
-	// UnsubscribeAll removes all listeners
+	// UnsubscribeAll close all listeners channels
 	UnsubscribeAll()
 }
 
-type eventSubscription struct {
-	Chan   chan Event
-	Cancel context.CancelFunc
-}
-
 // EventEmitter Registers listeners and dispatches events to them
 type EventEmitter struct {
-	Subscribers []*eventSubscription
-	lock        sync.RWMutex
+	subs []*eventSubscription
+	lock sync.Mutex
 }
 
-func (e *EventEmitter) allEventSubscription() []*eventSubscription {
-	e.lock.RLock()
-	defer e.lock.RUnlock()
+func (e *EventEmitter) UnsubscribeAll() {
+	for _, s := range e.allEventSubscription() {
+		s.cancel()
+	}
+}
 
-	return e.Subscribers
+type eventSubscription struct {
+	ch      chan Event
+	sm      *semaphore.Weighted
+	mu      sync.Mutex
+	closing sync.Once
+	cancel  context.CancelFunc
+	ctx     context.Context
 }
 
-func (e *EventEmitter) UnsubscribeAll() {
-	for _, c := range e.allEventSubscription() {
-		c.Cancel()
-	}
+func (e *EventEmitter) allEventSubscription() []*eventSubscription {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	return append([]*eventSubscription(nil), e.subs...)
 }
 
-func (e *EventEmitter) Emit(evt Event) {
-	for _, s := range e.allEventSubscription() {
-		select {
-		case s.Chan <- evt:
-			break
-		default:
-			break
-		}
+func (e *EventEmitter) Emit(ctx context.Context, evt Event) {
+	for _, sub := range e.allEventSubscription() {
+		go func(sub *eventSubscription) {
+			sub.mu.Lock()
+			defer sub.mu.Unlock()
+
+			if err := sub.sm.Acquire(sub.ctx, 1); err != nil {
+				return
+			}
+
+			sub.ch <- evt
+
+			sub.sm.Release(1)
+		}(sub)
 	}
 }
 
-func (e *EventEmitter) Subscribe(ctx context.Context, handler func(Event)) {
-	ctx, cancelFunc := context.WithCancel(ctx)
+func (s *eventSubscription) close() {
+	s.closing.Do(func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
 
-	ch := make(chan Event, 50)
+		_ = s.sm.Acquire(s.ctx, 1)
 
-	sub := &eventSubscription{
-		Chan:   ch,
-		Cancel: cancelFunc,
-	}
+		s.cancel()
+		close(s.ch)
+	})
+}
 
+func (e *EventEmitter) removeSub(s *eventSubscription) {
 	e.lock.Lock()
-	e.Subscribers = append(e.Subscribers, sub)
-	e.lock.Unlock()
+	defer e.lock.Unlock()
 
-	for {
-		select {
-		case <-ctx.Done():
-			e.unsubscribe(sub)
-			return
+	subs := e.subs
 
-		case evt := <-ch:
-			handler(evt)
+	for i, can := range subs {
+		if can == s {
+			subs[i] = subs[len(subs)-1]
+			e.subs = subs[:len(subs)-1]
+			break
```

Comment:
sorry I didn't run the code, but I have got the feeling that it won't just remove the subscription from the list but the subscription + all the following ones

can you double-check that it does what it should?

Reply:
Actually i move the last item from the list to the index of the removed one, it should be ok

https://github.com/golang/go/wiki/SliceTricks#delete-without-preserving-order

Reply:
I updated the code as the slice stores pointers to structs